core/src/test/java/org/testcontainers/junit/DependenciesTest.java

```diff
@@ -0,0 +1,116 @@
+package org.testcontainers.junit;
+
+import lombok.Getter;
+import org.junit.Test;
+import org.rnorth.visibleassertions.VisibleAssertions;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.startupcheck.OneShotStartupCheckStrategy;
+import org.testcontainers.lifecycle.Startable;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class DependenciesTest {
+
+    @Test
+    public void shouldWorkWithSimpleDependency() {
+        InvocationCountingStartable startable = new InvocationCountingStartable();
+
+        try (
+            GenericContainer container = new GenericContainer()
+                .withStartupCheckStrategy(new OneShotStartupCheckStrategy())
+                .dependsOn(startable)
+        ) {
+            container.start();
+        }
+
+        VisibleAssertions.assertEquals("Started once", 1, startable.getStartInvocationCount().intValue());
+        VisibleAssertions.assertEquals("Does not trigger .stop()", 0, startable.getStopInvocationCount().intValue());
+    }
+
+    @Test
+    public void shouldWorkWithMutlipleDependencies() {
+        InvocationCountingStartable startable1 = new InvocationCountingStartable();
+        InvocationCountingStartable startable2 = new InvocationCountingStartable();
+
+        try (
+            GenericContainer container = new GenericContainer()
+                .withStartupCheckStrategy(new OneShotStartupCheckStrategy())
+                .dependsOn(startable1, startable2)
+        ) {
+            container.start();
+        }
+
+        VisibleAssertions.assertEquals("Startable1 started once", 1, startable1.getStartInvocationCount().intValue());
+        VisibleAssertions.assertEquals("Startable2 started once", 1, startable2.getStartInvocationCount().intValue());
+    }
+
+    @Test
+    public void shouldStartEveryTime() {
+        InvocationCountingStartable startable = new InvocationCountingStartable();
+
+        try (
+            GenericContainer container = new GenericContainer()
+                .withStartupCheckStrategy(new OneShotStartupCheckStrategy())
+                .dependsOn(startable)
+        ) {
+            container.start();
+            container.stop();
+
+            container.start();
+            container.stop();
+
+            container.start();
+        }
+
+        VisibleAssertions.assertEquals("Started multiple times", 3, startable.getStartInvocationCount().intValue());
+        VisibleAssertions.assertEquals("Does not trigger .stop()", 0, startable.getStopInvocationCount().intValue());
+    }
+
+    @Test
+    public void shouldStartTransitiveDependencies() {
```

Comment:
Just eyeballing the code, it looks like we're safe from circular dependencies, and also 'diamond' dependencies should be OK too.

Do you think it might be good to have tests for these scenarios in case a regression is introduced later?

Reply:
Circular dependencies won't work (we have "hard depends on")

I will add a test for the diamond case 👍